pandas/_libs/lib.pyx

```diff
@@ -2821,7 +2829,7 @@ NoDefault = Literal[_NoDefault.no_default]
 
 
 def map_infer_mask(
-        ndarray[object] arr,
+        ndarray arr,
```

Comment:
This will make iteration slower.

Look into ``PyArray_GETITEM``/``PyArray_ITER_NEXT`` like in ``Validator._validate``?

Maybe this can be a separate PR?